Notifications: improve Dynamic Type support in the body of the notification

jaclync opened the pull request:
Fixes #912  

## Changes
The fixes in this PR are for two main issues:
- The font size of the subject and snippet labels did not update dynamically, but only on app launch:
  -  I haven't found concrete/official evidence online to support my fix, but I believe it's caused by using a static constant `NSAttributedString` styles dictionary with a fixed [`preferredFont(forTextStyle:)`](https://developer.apple.com/documentation/uikit/uifont/1619030-preferredfont) value for all `NSAttributedString` usage. From all the official and popular examples using `preferredFont(forTextStyle:)` for dynamic type, it is always called on `font` assignment.Open to any other thoughts!
- The notification icon on the left is clipped with larger font sizes. I fixed this one by:
  - Setting `horizontalCompressionResistancePriority` to `required` so that the icon maintains its content size with the highest priority
  - Creating a `UIFont` helper to return a `Noticons` font that supports dynamic type based on a font size, following [Apple documentation](https://developer.apple.com/documentation/uikit/uifont/scaling_fonts_automatically) on custom font. I originally wanted to update the [existing `Noticons` font helper](https://github.com/woocommerce/woocommerce-ios/blob/develop/WooCommerce/Classes/Extensions/UIFont+Helpers.swift#L59-L63), which calculates the font size at the given device font size and returns a `UIFont` with that fixed calculated font size. Because the existing helper works differently (the return `UIFont` does not scale with dynamic type) and is used in other cases, I decided to create a separate helper
  - Notes: I kept the text style for the icon as `title1`, but I also tried `body` and it looks slightly more aligned with the subject with larger font sizes (as shown in screenshots). The icon base font size can also be tweaked - see if @kyleaparker has any preferences. --> Updated to `body` font style

## Testing
- Launch the app with a store that has at least one notification
- Go to "Notifications" tab
- Turn on dynamic type (device Settings > General > Accessibility > Larger Text > turn it on) and adjust the device to different font sizes via the slider, and observe the notification cell --> The labels should adjust their font size dynamically

## Example screenshots

Notes: the notification content and icon color are mocked

text style | 3rd smallest | 8th smallest | largest
--- | --- | --- | ---
`body` with 25pt base font size | ![Simulator Screen Shot - iPhone SE - 2019-07-08 at 17 39 52](https://user-images.githubusercontent.com/1945542/60844532-8c7b1e00-a1a7-11e9-9024-e23ac68e63eb.png) | ![Simulator Screen Shot - iPhone SE - 2019-07-08 at 17 40 01](https://user-images.githubusercontent.com/1945542/60844538-91d86880-a1a7-11e9-9c36-dc866aaed21a.png) | ![Simulator Screen Shot - iPhone SE - 2019-07-08 at 17 39 45](https://user-images.githubusercontent.com/1945542/60844542-943ac280-a1a7-11e9-881f-dec5fff22fc1.png)




Update release notes:

- [ ] If there are user facing changes, I have added an item to `RELEASE-NOTES.txt`.

mindgraffiti commented on WooCommerce/Classes/ViewRelated/Notifications/Cells/NoteTableViewCell.swift:
```diff
@@ -98,7 +98,7 @@ class NoteTableViewCell: UITableViewCell {
     override func awakeFromNib() {
         super.awakeFromNib()
-        noticonLabel.font = UIFont.noticon(forStyle: .title1)
+        noticonLabel.font = UIFont.noticon(forStyle: .title1, baseSize: 25.0)
```
I know this wasn't your code. Please change from `title1` to `body`

mindgraffiti commented:
No, that's part of Apple's defaults. If you go to your Settings app > General > Accessibility > Larger Text and turn on the larger text then swipe to the largest text setting (on the bottom selection thingy), you will see the exact same thing happen to the section headers in the settings app.

jaclync commented:
@mindgraffiti thanks a lot for explaining where to find the design files! I got to download the Sketch file, but I couldn't find where the font was specified (the Noticons glyphs seem to be symbols in Sketch):
<img width="889" alt="Screen Shot 2019-07-08 at 4 51 43 PM" src="https://user-images.githubusercontent.com/1945542/60841928-f0e6af00-a1a0-11e9-8ae5-c24caa766f21.png">
not sure if there is another way to check on the font?

> I noticed your PR does not fix the truncated text for the notification's content text. (Example: "New order for $100.00 on steep and sip")

great catch! When I looked at the screenshot in the issue, the obvious truncated UI was the glyph icon, since I'm used to the message in iOS notifications truncated after like 4 lines. Will fix the truncated text in this PR!

mindgraffiti commented:
>I couldn't find where the font was specified (the Noticons glyphs seem to be symbols in Sketch)

Yep, that's true. The fonts will only be useful for text. Any icons you see should be found in the Gridicons CocoaPod https://github.com/Automattic/Gridicons-iOS. 

>Will fix the truncated text in this PR!

Sounds good 😀

jaclync commented:
just updated to use body font style for the Noticon icon, and fix truncated text in the notification content! (screenshots in PR description also updated)

mindgraffiti commented on WooCommerce/Classes/ViewRelated/Notifications/Cells/NoteTableViewCell.xib:
```diff
@@ -31,13 +31,13 @@
                         <rect key="frame" x="55" y="11" width="304" height="56.5"/>
                         <subviews>
-                            <label opaque="NO" userInteractionEnabled="NO" contentMode="left" text="Subject" textAlignment="natural" lineBreakMode="tailTruncation" numberOfLines="2" baselineAdjustment="alignBaselines" adjustsFontForContentSizeCategory="YES" adjustsFontSizeToFit="NO" translatesAutoresizingMaskIntoConstraints="NO" id="Ow4-CR-HAI" userLabel="Subject Label">
+                            <label opaque="NO" userInteractionEnabled="NO" contentMode="left" text="Subject" textAlignment="natural" lineBreakMode="tailTruncation" numberOfLines="0" baselineAdjustment="alignBaselines" adjustsFontForContentSizeCategory="YES" adjustsFontSizeToFit="NO" translatesAutoresizingMaskIntoConstraints="NO" id="Ow4-CR-HAI" userLabel="Subject Label">
```
`lineBreakMode` needs set to `Word Wrap` in order for dynamic text changes to register.

mindgraffiti commented on WooCommerce/Classes/ViewRelated/Notifications/Cells/NoteTableViewCell.xib:
```diff
@@ -31,13 +31,13 @@
                                 <nil key="highlightedColor"/>
                             </label>
-                            <label opaque="NO" userInteractionEnabled="NO" contentMode="left" horizontalHuggingPriority="251" verticalHuggingPriority="251" horizontalCompressionResistancePriority="752" text="Snippet" textAlignment="natural" lineBreakMode="tailTruncation" numberOfLines="2" baselineAdjustment="alignBaselines" adjustsFontForContentSizeCategory="YES" adjustsFontSizeToFit="NO" transla ...
+                            <label opaque="NO" userInteractionEnabled="NO" contentMode="left" horizontalHuggingPriority="251" verticalHuggingPriority="251" horizontalCompressionResistancePriority="752" text="Snippet" textAlignment="natural" lineBreakMode="tailTruncation" numberOfLines="0" baselineAdjustment="alignBaselines" adjustsFontForContentSizeCategory="YES" adjustsFontSizeToFit="NO" transla ...
```
Line Break = Word Wrap here as well.

jaclync commented:
> The Subject and Snippet labels were not resizing when tested on a device. Change the Line Break mode to be Word Wrap for both the labels, so that it resizes.

What type of device did you use where you saw the subject/snippet labels not resizing with device font size? It worked on my simulator and device (iPhone XR) without the `lineBreakMode` change, but I just added a commit to update `lineBreakMode` to `wordWrap` for your testing. Not sure why it would vary by device, sounds a bit concerning :/

> Add Release notes explaining the bug fix

Hmm from my understanding of the checkbox in PR template, only PRs with user-facing features go into the release notes. Maybe I misunderstood and all UI bug fixes should be added? If that's the case, I'll open a separate PR to add previous UI bugs that I didn't list in the release notes 😓

mindgraffiti commented:
>Hmm from my understanding of the checkbox in PR template, only PRs with user-facing features go into the release notes. 

That's interesting. It sounds like we'll need to update the release notes checkbox to be a bit clearer.

We've been putting all user-facing features and any notable bug fixes in the release notes. Accessibility is something that @ctarda and I care a LOT about, so anytime we do something to make our accessibility better, we mention it.

mindgraffiti commented:
My device is an iPhone XS, with iOS 12.3.1 running. I could also reproduce the problem on the simulator, running the iPhone XS with iOS 12.2.

jaclync commented:
> The Subject and Snippet labels were not resizing when tested on a device.

Hmm that's interesting my device has the same iOS version. Just double checking, did you see the truncated text in this commit https://github.com/woocommerce/woocommerce-ios/pull/1068/commits/9e65a5098fac55153b32bfa4df05959510504745 but not in the latest https://github.com/woocommerce/woocommerce-ios/pull/1068/commits/29359c6a1b9a13efe6df119bd381cc8f510a8da9 ?

> Add Release notes explaining the bug fix

Yep I totally value any improvements on accessibility, just not super sure how the bug fixes are translated into App Store release notes - via PR labels/description/title or developer's edits of release notes. Just to clarify, do we add all non-UI/UI bug fixes and features into the release notes, or just UI/user-facing bug fixes and features?

mindgraffiti commented:
This is how Release Notes cycle is handled p77Llu-bGf-p2

>Just to clarify, do we add all non-UI/UI bug fixes and features into the release notes, or just UI/user-facing bug fixes and features?

It's not formalized. As a general rule, trust your own good judgement 😀

Generally, we add all user-facing features and any notable bug fixes. Not all bug fixes that get listed in the release notes are UI-based. If you'd prefer more clarity on it, feel free to propose some new rules. We're open to suggestions!

Testing out your individual commits now. Will remember to post screenshots this time :)

mindgraffiti commented:
Re-tested on my device and the sim. Realized I had the wrong branch checked out the whole time 🤦‍♀  

It works perfectly with commit 9e65a50. Thanks for being so patient! :shipit:

jaclync commented:
Hah no problem, it happens esp when switching branches a lot! I'll revert https://github.com/woocommerce/woocommerce-ios/pull/1068/commits/29359c6a1b9a13efe6df119bd381cc8f510a8da9 and merge since we both tested on the previous commit https://github.com/woocommerce/woocommerce-ios/commit/9e65a5098fac55153b32bfa4df05959510504745, thanks for testing & the suggestions!
